benchmark/src/main/java/com/google/jetpackcamera/benchmark/ImageCaptureLatencyBenchmark.kt

```diff
@@ -0,0 +1,103 @@
+import androidx.benchmark.macro.StartupMode
+import androidx.benchmark.macro.TraceSectionMetric
+import androidx.benchmark.macro.junit4.MacrobenchmarkRule
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import androidx.test.uiautomator.By
+import androidx.test.uiautomator.Until
+import org.junit.Assert.fail
+import org.junit.Rule
+import org.junit.Test
+import org.junit.runner.RunWith
+
+@RunWith(AndroidJUnit4::class)
+class ImageCaptureLatencyBenchmark {
+    @get:Rule
+    val benchmarkRule = MacrobenchmarkRule()
+
+    @Test
+    fun rearCameraNoFlashLatency() {
+        imageCaptureLatency(shouldFaceFront = false, flashMode = FlashMode.OFF)
+    }
+
+    @Test
+    fun frontCameraNoFlashLatency() {
+        imageCaptureLatency(shouldFaceFront = true, flashMode = FlashMode.OFF)
+    }
+
+    @Test
+    fun rearCameraWithFlashLatency() {
+        // Flash test needs extra time at the end to ensure the trace is closed
+        imageCaptureLatency(shouldFaceFront = false, flashMode = FlashMode.ON, sleepInterval = 5000)
+    }
+
+    // todo(kimblebee): front flash latency test
+
+    /**
+     * Measures the time between an onClick event on the Capture Button and onImageCapture
+     * callback being fired from
+     * [takePicture][com.google.jetpackcamera.domain.camera.CameraXCameraUseCase.takePicture].
+     *
+     *  @param shouldFaceFront the direction the camera should be facing.
+     *  @param flashMode the designated [FlashMode] for the camera.
+     *  @param sleepInterval option to change the default sleep interval after performing clicking
+     *  the Image Capture button.
+     *
+     */
+    @OptIn(ExperimentalMetricApi::class)
+    private fun imageCaptureLatency(
+        shouldFaceFront: Boolean,
+        flashMode: FlashMode,
+        sleepInterval: Long = 500
+    ) {
+        benchmarkRule.measureRepeated(
+            packageName = JCA_PACKAGE_NAME,
+            metrics = listOf(
+                TraceSectionMetric(sectionName = IMAGE_CAPTURE_TRACE, targetPackageOnly = false)
+            ),
+            iterations = DEFAULT_TEST_ITERATIONS,
+            startupMode = StartupMode.WARM,
+            setupBlock = {
+                allowCamera()
+                pressHome()
+                startActivityAndWait()
+                toggleQuickSettings(device)
+                setQuickFrontFacingCamera(shouldFaceFront = shouldFaceFront, device = device)
+                setQuickSetFlash(flashMode = flashMode, device = device)
+
+                toggleQuickSettings(device)
+                device.waitForIdle()
+            }
+
+        ) {
+            val selector = By.res("CaptureButton")
+
+            if (!device.wait(Until.hasObject(selector), 2_500)) {
+                fail("Did not find object with id CaptureButton")
+            }
+            device
+                .findObject(selector)
+                .click()
+
+            // ensure trace is closed
+            Thread.sleep(sleepInterval)
```

Comment:
There are 2 things that waiting for a real tag accomplishes:
1. It ensures that the capture has finished, and therefore the trace metric you're capturing will be fully completed.
2. It ensures that when the capture completes successfully, the test will also finish shortly after.

Waiting for an element that will never arrive is basically equivalent to `Thread.sleep`, which doesn't work for either of the above cases:
1. No matter what `sleepInterval` you set, you can never choose a length of time that will guarantee the capture has finished. This is what leads to flaky test results since we may get tests that "succeed", but the trace we're monitoring still hasn't finished.
2. If we wait for the full `sleepInterval`, a test that may have taken 100ms will now take a few seconds. This may not seem like a big deal, but when the codebase includes 100 tests that take 2-3 seconds longer than they should, now running all of the tests takes a few minutes longer than needed. Those few minutes add up when you are running the tests constantly in CI or locally when you're trying to make a change.

The reason I mentioned adding a much longer timeout when waiting for a real element is because timeouts should be treated as test failures. We should choose the timeout to be a time interval where we feel like something has gone so wrong that capturing the benchmark results probably won't be useful. When that timeout hits, it will throw an exception that will signal to macrobenchmark to stop collecting metrics. Something like 10-15 seconds will probably be sufficient in this case, but we should also monitor the metrics collected and make adjustments to that if needed.

Reply:
oh ok that makes a lot of sense.
However, I do wonder one more thing. Eventually a post capture screen will be integrated into the app, and i'm not entirely sure if it will need to keep the toast notification (since the presence post capture screen itself should indicate a successful capture). 
Should I wait for Tahsin's PR to merge first, then this one? or i could also create a toast element to target with the intention of eventually being removed after post capture integration

Reply:
Yes, we'll need to change these tests once we no longer show the Toast. Luckily, the tests will start failing if we get rid of the Toast, so it will be obvious what needs to be changed.

Adding a Toast before Tahsin's PR is merged is fine. If you want to do that, I would recommend just copying the logic from that PR into this PR (or a separate PR) so there aren't complex merge conflicts when one or the other PR is merged.